server/bytestream/zip.go

```diff
@@ -0,0 +1,190 @@
+// This file is primarily adapted from useful (unexported) functions from
+// the golang package 'archive/zip', slightly modified to match our use
+// cases.
+
+// Copyright (c) 2009 The Go Authors. All rights reserved.
+
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+
+//    * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//    * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//    * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package bytestream
+
+import (
+	"encoding/binary"
+	"errors"
+
+	arpb "github.com/buildbuddy-io/buildbuddy/proto/archive"
+)
+
+const (
+	fileHeaderSignature      = 0x04034b50
+	directoryHeaderSignature = 0x02014b50
+	directoryEndLen          = 22
+	directoryHeaderLen       = 46
+	fileHeaderLen            = 30
+)
+
+var (
+	ErrFormat    = errors.New("zip: not a valid zip file")
+	ErrAlgorithm = errors.New("zip: unsupported compression algorithm")
+	ErrChecksum  = errors.New("zip: checksum error")
+	ErrZip64     = errors.New("zip: zip64 not supported")
```

Comment:
we don't usually use errors.New because they don't go over the wire very well. You can grab an appropriate error type from server/util/status and then it will be a proper gRPC error if it ever goes across the wire (and we'll be able to parse it etc).

Also not 100% sure you want these to be exported from this package but maybe?

Reply:
Yeah, this were public in the original package, but I made them private now and updated to hopefully reasonable error types.

I have a few other IO-related errors (theoretical-but-would-be-weird error outputs from writes to Writers, etc.) that are new and look like they might leak out.  Should I be wrapping those up as well?